File: mergesort.hs
merge :: (a -> a -> Bool) -> [a] -> [a] -> [a]
merge _      x     [] = x
merge _     []      y = y
merge f (x:xs) (y:ys)
                    | f x y     = y:(merge (x:xs) ys)
                    | otherwise = x:(merge xs (y:ys))

mergesort :: (a -> a -> Bool) -> [a] -> [a] 
mergesort _     [] = []
mergesort _    [x] = [x] 
mergesort f [x, y]
                 | f x y     = [y, x]
                 | otherwise = [x, y]
mergesort f x = merge f (mergesort f (take p x)) (mergesort f (drop p x))
                where
                    p = div (length x) 2

main :: IO()
main = print $ mergesort (>) [5,9,1,3,4,6,6,3,2]

File: mergesort.exs
defmodule Mergesort do
  def merge([], y), do: y
  def merge(x, []), do: x

  def merge([x | xs], [y | ys]) do
    if x > y do
      [y | merge([x | xs], ys)]
    else
      [x | merge(xs, [y | ys])]
    end
  end

  def mergesort([]), do: []
  def mergesort([x | []]), do: [x]
  def mergesort([x, y | []]), do: if(x > y, do: [y, x], else: [x, y])

  def mergesort(l) do
    p = div(length(l), 2)
    merge(mergesort(Enum.take(l, p)), mergesort(Enum.drop(l, p)))
  end
end

IO.inspect(Mergesort.mergesort([5, 9, 1, 3, 4, 6, 6, 3, 2]))
